File: zadoff_chu_sequence.py
import cmath
import math
import numpy as np

def print_sequence(symbols):
    for i, symbol in enumerate(symbols, start=1):
        # Ensure symbol is a complex number
        symbol = complex(symbol.real, symbol.imag)
        r = abs(symbol)
        theta = cmath.phase(symbol)
        print(f'#{i:02d}  {symbol.real:+.4f} {symbol.imag:+.4f}j, r = {r:.1f}, theta = {theta:+.4f} = {math.degrees(theta):+.1f}')

def zadoff_chu_sequence(n_zc, u, q=0):
    """
    Generates a Zadoff-Chu sequence.

    Parameters:
    - n_zc (int): Length of the Zadoff-Chu sequence (must be a positive integer).
    - u (int): Root index of the Zadoff-Chu sequence (must satisfy 0 < u < n_zc and be coprime with n_zc).
    - q (int, optional): Cyclic shift parameter (default is 0).

    Returns:
    - numpy.ndarray: The generated Zadoff-Chu sequence as a NumPy array of complex numbers.

    Raises:
    - ValueError: If input parameters do not meet the required conditions.
    """

    # Validate input parameters
    if n_zc <= 0:
        raise ValueError('"n_zc" must be a positive integer.')
    if u <= 0 or u >= n_zc:
        raise ValueError('"u" must satisfy 0 < u < n_zc.')
    if math.gcd(n_zc, u) != 1:
        raise ValueError('"u" and "n_zc" must be coprime (gcd(n_zc, u) == 1).')

    # Determine whether n_zc is even or odd (cf = conjugate flag)
    cf = n_zc % 2

    # Generate sequence indices
    n = np.arange(n_zc)

    # Calculate the exponent for the Zadoff-Chu sequence
    exponent = -1j * np.pi * u * n * (n + cf + 2 * q) / n_zc

    # Generate the Zadoff-Chu sequence
    sequence = np.exp(exponent)

    return sequence


if __name__ == "__main__":
    zc_sequence = zadoff_chu_sequence(10, 7, 1)
    print_sequence(zc_sequence)

#01  +1.0000 +0.0000j, r = 1.0, theta = +0.0000 = +0.0
#02  +0.9511 -0.3090j, r = 1.0, theta = -0.3142 = -18.0
#03  +0.3090 +0.9511j, r = 1.0, theta = +1.2566 = +72.0
#04  -0.0000 -1.0000j, r = 1.0, theta = -1.5708 = -90.0
#05  -0.8090 -0.5878j, r = 1.0, theta = -2.5133 = -144.0
#06  -0.0000 -1.0000j, r = 1.0, theta = -1.5708 = -90.0
#07  +0.3090 +0.9511j, r = 1.0, theta = +1.2566 = +72.0
#08  +0.9511 -0.3090j, r = 1.0, theta = -0.3142 = -18.0
#09  +1.0000 +0.0000j, r = 1.0, theta = +0.0000 = +0.0
#10  -0.5878 +0.8090j, r = 1.0, theta = +2.1991 = +126.0

File: zadoff_chu_sequence.m
% 
% zadoff_chu_sequence() Generate a Zadoff-Chu sequence.
% 
function y = zadoff_chu_sequence(N, R, Q)
    %
    %   y = zadoff_chu_sequence(N, R, Q) returns a Zadoff-Chu sequence of length N,
    %   with root index R, and cyclic shift Q.
    %
    %   The sequence is defined as:
    %       y(n) = exp(-1j * pi * R * n * (n + cf + 2 * Q) / N)
    %   where n = 0, 1, ..., N-1, and cf = mod(N, 2).
    %
    %   Inputs:
    %       N - Sequence length (positive integer)
    %       R - Root index (integer satisfying gcd(N, R) == 1)
    %       Q - Cyclic shift (integer, default is 0)
    %
    %   Output:
    %       y - Generated Zadoff-Chu sequence (complex column vector)
    %
    %   Example:
    %       N = 63;
    %       R = 5;
    %       Q = 0;
    %       y = zadoff_chu_sequence(N, R, Q);
    %
    %   References:
    %       3GPP TS 36.211 - Physical Channels and Modulation
    
    % Set default value for Q if not provided
    if nargin < 3
        Q = 0;
    end

    % Input validation
    if ~isscalar(N) || ~isnumeric(N) || N <= 0 || floor(N) ~= N
        error('N must be a positive integer.');
    end

    if ~isscalar(R) || ~isnumeric(R) || floor(R) ~= R
        error('R must be an integer.');
    end

    if ~isscalar(Q) || ~isnumeric(Q) || floor(Q) ~= Q
        error('Q must be an integer.');
    end

    % Check if N and R are coprime
    if gcd(N, R) ~= 1
        error('ZC sequence length N and parameter R must be coprime. %d and %d are not coprime.', N, R);
    end

    % Determine whether N is even or odd (cf = conjugate flag)
    cf = mod(N, 2); % cf = 0 for even N, cf = 1 for odd N

    % Generate sequence indices
    n = (0:N-1).'; % Column vector for consistency

    % Calculate the exponent for the Zadoff-Chu sequence
    exponent = -1j * pi * R * n .* (n + cf + 2 * Q) / N;

    % Generate the Zadoff-Chu sequence
    y = exp(exponent);
end
    